### PR TITLE
Feature/choice vr optional

### LINKED ISSUE #2160: [SDL 0064] Choice-VR optional
Proposal: [Choice-VR optional](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0064-choice-vr-optional.md)

> This proposal is about changing Choice.vrCommands to be not mandatory.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/189

Steering Committee Decision:

> The Steering Committee agreed to accept this proposal and also agreed that the error for mixed vrCommand should be more descriptive, like info = "Some choices don't contain VR commands. Either all or none must have voice commands."

### PR DESCRIPTION
**Pull request by conlain-k**
Fixes #2160

This PR is **ready** for review.

### Risk
This PR makes **major** API changes.

### Testing Plan
add unit and ATF tests, ensure no breakage in old tests (except the ones testing removed functionality)

### Summary
Make the Choice.vrCommands parameter optional and require choice sets to be consistent in whether their choices have the parameter. Allow PerformInteraction requests in MANUAL_ONLY mode to have choices without vrCommands.

##### Breaking Changes
* makes an optional parameter mandatory

### Tasks Remaining:
- [x] Cleanup and finish development
- [ ] add tests and check old ones for regressions

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/application_manager/include/application_manager/message_helper.h [hunk 1]
```diff
@@ -855,6 +856,43 @@ class MessageHelper {
   static smart_objects::SmartObjectSPtr CreateMessageForHMI(
       hmi_apis::messageType::eType message_type, const uint32_t correlation_id);
 
+  // Check whether each choice has the vrCommands field
+  // returns -1 for failure, 0 if all choice include vrCommands, and 1 if none
+  // do
+  // vrCommands is an all-or-none deal
+  static int CheckChoiceSet_VRCommands(
+      const smart_objects::SmartObject& choice_set) {
+    // if this becomes true, someone doesn't have vrCommands
```
Comment (jacobkeeler) [inline src/components/application_manager/include/application_manager/message_helper.h]: `becomes false`?

File: src/components/application_manager/include/application_manager/message_helper.h [hunk 2]
```diff
@@ -855,6 +856,43 @@ class MessageHelper {
   static smart_objects::SmartObjectSPtr CreateMessageForHMI(
       hmi_apis::messageType::eType message_type, const uint32_t correlation_id);
 
+  // Check whether each choice has the vrCommands field
+  // returns -1 for failure, 0 if all choice include vrCommands, and 1 if none
+  // do
+  // vrCommands is an all-or-none deal
```
Comment (jacobkeeler) [inline src/components/application_manager/include/application_manager/message_helper.h]: Should do a proper doxygen header

File: src/components/application_manager/include/application_manager/message_helper.h [hunk 3]
```diff
@@ -855,6 +856,43 @@ class MessageHelper {
...
+    // everyone has it
+    if (all_have) {
+      return 0;
+    }
+    // No one has it
+    if (none_have) {
+      return 1;
+    }
+    // mix-and-match, this is an error
+    return -1;
```
Comment (jacobkeeler) [inline src/components/application_manager/include/application_manager/message_helper.h]: Implementation should be implementation file

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/perform_interaction_request.h [hunk 4]
```diff
@@ -212,6 +212,16 @@ class PerformInteractionRequest
...
 
+  /**
+   * @brief Checks each choice in each set for having a VRcommands parameter
+   * @param app contains pointer to application.
+   * @param choice_set_id_list_length contains amount
+   * of choice set ids.
+   * @param choice_set_id_list array of choice set ids
+   * @return returns false request has choice sets with no vrCommands
+   */
+  bool CheckChoiceSetList_VRCommands(app_mngr::ApplicationSharedPtr app);
```
Comment (jacobkeeler) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/perform_interaction_request.h]: Same issue, no underscores.

**Comment by conlain-k** [comment]
should also be able to close #2400 and #2401 with this

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/create_interaction_choice_set_test.cc [hunk 5]
```diff
@@ -481,6 +487,40 @@ TEST_F(CreateInteractionChoiceSetRequestTest,
   command_->Run();
 }
 
+TEST_F(CreateInteractionChoiceSetRequestTest,
+       Run_WithoutVrCommands_SUCCESS) {
```
Comment (jacobkeeler) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/create_interaction_choice_set_test.cc]: Should probably have a test case for MIXED as well.

**Comment by LuxoftAKutsan** [comment]
@conlain-k Looks like style is broken on the PR http://opensdl-jenkins.luxoft.com:8080/job/OpenSDL_PR/1907/console

**Pull request merged**